### ISSUE TITLE
Deferred route drawing causes problems with some tests

### ISSUE THREAD

**Issue by rosa** [rails/rails#51906]
This issue is related to https://github.com/rails/rails/pull/51614

### Steps to reproduce
I haven't been able to write a minimal reproduction script. In our case, the issue we had was that a test for a helper that used some URL helpers and these are undefined after #51614. This is how the test looks like: 

```ruby
require "test_helper"

class AttachmentsBrowserHelperTest < ActionView::TestCase
  test "attachments_browser_backlink with a context" do
    assert_match /Back to Sent Mail/, attachments_browser_backlink(context: topics_sent_path)
  end
```

Looks like #51614 caused some issues with `rspec-rails` as well, according to [this comment](https://github.com/rails/rails/pull/51614#issuecomment-2123136767) and https://github.com/rspec/rspec-rails/issues/2761. 

### Expected behavior
The test passes normally. 

### Actual behavior
It fails with 
```
NameError: undefined local variable or method `topics_sent_path' for an instance of AttachmentsBrowserHelperTest
    test/helpers/attachments_browser_helper_test.rb:17:in `block in <class:AttachmentsBrowserHelperTest>'
```

To work around this, I've used the following: 
```ruby
 setup do
    routes_reloader = Rails.application.routes_reloader
    routes_reloader.execute_unless_loaded if routes_reloader&.respond_to?(:execute_unless_loaded)
  end
```

### System configuration
**Rails version**: Edge (targeting main branch)

**Ruby version**: 3.3.0
